File: client/services/ConnectionPool.ts
// =============================================================================
// WebSocket Engine - Connection Pool
// =============================================================================

import { logger } from '../utils/logger';

interface EndpointHealth {
  url: string;
  score: number;
  lastSuccess: number | null;
  lastFailure: number | null;
  consecutiveFailures: number;
  consecutiveSuccesses: number;
  averageLatency: number;
  latencyHistory: number[];
  totalRequests: number;
  failedRequests: number;
  lastChecked: number;
}

interface ConnectionMetrics {
  endpoint: string;
  connectedAt: number;
  messagesReceived: number;
  messagesSent: number;
  bytesReceived: number;
  bytesSent: number;
  lastActivity: number;
}

export interface ConnectionPoolConfig {
  maxPerEndpoint?: number;
  healthCheckInterval?: number;
  latencyHistorySize?: number;
  scoreDecayRate?: number;
  minHealthScore?: number;
  loadBalancingStrategy?: 'weighted-random' | 'least-connections' | 'round-robin';
  healthCheckTimeout?: number;
  enableHealthCheck?: boolean;
  endpoints?: string[];
}

export class ConnectionPool {
  private endpoints: Map<string, EndpointHealth> = new Map();
  private connections: Map<string, WebSocket[]> = new Map();
  private connectionMetrics: Map<string, ConnectionMetrics> = new Map();
  private healthScores: Map<string, number> = new Map();
  private activeEndpoint: string | null = null;
  private preferredEndpoint: string | null = null;
  private isDestroyed = false;

  private config = {
    maxPerEndpoint: 3,
    healthCheckInterval: 30000,
    latencyHistorySize: 100,
    scoreDecayRate: 0.1,
    minHealthScore: 10,
    loadBalancingStrategy: 'weighted-random' as 'weighted-random' | 'least-connections' | 'round-robin',
    healthCheckTimeout: 5000,
    enableHealthCheck: true,
  };

  private healthCheckTimer: NodeJS.Timeout | null = null;
  private roundRobinIndex = 0;
  private healthCheckInProgress = false;

  constructor(config?: ConnectionPoolConfig) {
    Object.assign(this.config, config);

    const endpoints = config?.endpoints || [];
    const endpointStrings: string[] = Array.isArray(endpoints)
      ? endpoints.filter((e): e is string => typeof e === 'string')
      : [];

    const uniqueEndpoints = Array.from(new Set(endpointStrings));

    uniqueEndpoints.forEach((endpoint) => {
      this.addEndpoint(endpoint);
    });

    if (this.config.enableHealthCheck && this.config.healthCheckInterval > 0 && uniqueEndpoints.length > 0) {
      this.startHealthMonitoring();
    }
  }

  getBestEndpoint(): string {
    if (this.isDestroyed) {
      throw new Error('ConnectionPool has been destroyed');
    }

    const healthyEndpoints = this.getHealthyEndpoints();

    if (healthyEndpoints.length === 0) {
      const allEndpoints = Array.from(this.endpoints.entries());
      if (allEndpoints.length === 0) {
        throw new Error('No endpoints available');
      }
      allEndpoints.sort((a, b) => b[1].score - a[1].score);
      return allEndpoints[0]?.[0];
    }

    if (this.preferredEndpoint) {
      const preferredHealth = this.endpoints.get(this.preferredEndpoint);
      if (preferredHealth && preferredHealth.score > this.config.minHealthScore) {
        return this.preferredEndpoint;
      }
    }

    switch (this.config.loadBalancingStrategy) {
      case 'weighted-random':
        return this.selectWeightedRandom(healthyEndpoints);
      case 'least-connections':
        return this.selectLeastConnections(healthyEndpoints);
      case 'round-robin':
        return this.selectRoundRobin(healthyEndpoints);
      default:
        return healthyEndpoints[0][0];
    }
  }

  recordSuccess(endpoint: string, latency?: number): void {
    if (this.isDestroyed) return;

    const health = this.endpoints.get(endpoint);
    if (!health) return;

    health.lastSuccess = Date.now();
    health.consecutiveSuccesses++;
    health.consecutiveFailures = 0;
    health.totalRequests++;

    if (latency !== undefined && latency >= 0) {
      health.latencyHistory.push(latency);
      if (health.latencyHistory.length > this.config.latencyHistorySize) {
        health.latencyHistory.shift();
      }
      health.averageLatency = this.calculateAverageLatency(health.latencyHistory);
    }

    this.updateHealthScore(endpoint);
    this.healthScores.set(endpoint, health.score);
  }

  recordFailure(endpoint: string): void {
    if (this.isDestroyed) return;

    const health = this.endpoints.get(endpoint);
    if (!health) return;

    health.lastFailure = Date.now();
    health.consecutiveFailures++;
    health.consecutiveSuccesses = 0;
    health.totalRequests++;
    health.failedRequests++;

    this.updateHealthScore(endpoint, true);
    this.healthScores.set(endpoint, health.score);

    logger.warn(`Recorded failure for ${endpoint}, score: ${health.score}`);
  }

  addConnection(endpoint: string, ws: WebSocket): void {
    if (this.isDestroyed) return;

    if (!this.connections.has(endpoint)) {
      this.connections.set(endpoint, []);
    }

    const connections = this.connections.get(endpoint)!;

    if (connections.includes(ws)) {
      return;
    }

    if (connections.length < this.config.maxPerEndpoint) {
      connections.push(ws);
      logger.info(`Added connection to ${endpoint}, total: ${connections.length}`);
    } else {
      logger.warn(`Max connections (${this.config.maxPerEndpoint}) reached for ${endpoint}`);
    }
  }

  removeConnection(endpoint: string, ws: WebSocket): void {
    if (this.isDestroyed) return;

    const connections = this.connections.get(endpoint);
    if (connections) {
      const index = connections.indexOf(ws);
      if (index >= 0) {
        connections.splice(index, 1);
        logger.info(`Removed connection from ${endpoint}, remaining: ${connections.length}`);
      }
    }
  }

  getActiveConnection(): WebSocket | null {
    if (this.isDestroyed) return null;

    if (this.activeEndpoint) {
      const connections = this.connections.get(this.activeEndpoint);
      if (connections && connections.length > 0) {
        return connections.find(ws => ws.readyState === WebSocket.OPEN) || null;
      }
    }
    return null;
  }

  setActiveEndpoint(endpoint: string): void {
    if (!this.isDestroyed) {
      this.activeEndpoint = endpoint;
    }
  }

  setPreferredEndpoint(endpoint: string): void {
    if (!this.isDestroyed) {
      this.preferredEndpoint = endpoint;
    }
  }

  hasEndpoint(endpoint: string): boolean {
    return this.endpoints.has(endpoint);
  }

  addEndpoint(endpoint: string): void {
    if (this.isDestroyed) return;

    if (!this.endpoints.has(endpoint)) {
      this.endpoints.set(endpoint, {
        url: endpoint,
        score: 100,
        lastSuccess: null,
        lastFailure: null,
        consecutiveFailures: 0,
        consecutiveSuccesses: 0,
        averageLatency: 0,
        latencyHistory: [],
        totalRequests: 0,
        failedRequests: 0,
        lastChecked: Date.now(),
      });

      this.healthScores.set(endpoint, 100);
      this.connections.set(endpoint, []);

      logger.info(`Added endpoint: ${endpoint}`);

      if (this.config.enableHealthCheck && !this.healthCheckTimer && this.endpoints.size > 0) {
        this.startHealthMonitoring();
      }
    }
  }

  getHealthScores(): Record<string, number> {
    const scores: Record<string, number> = {};
    this.endpoints.forEach((health, endpoint) => {
      scores[endpoint] = health.score;
    });
    return scores;
  }

  getEndpointStats(): Array<{
    url: string;
    healthScore: number;
    connectionCount: number;
    isActive: boolean;
    isPreferred: boolean;
  }> {
    const stats = [];

    for (const [endpoint, health] of this.endpoints) {
      const connections = this.connections.get(endpoint) || [];

      stats.push({
        url: endpoint,
        healthScore: health.score,
        connectionCount: connections.filter(ws => ws.readyState === WebSocket.OPEN).length,
        isActive: endpoint === this.activeEndpoint,
        isPreferred: endpoint === this.preferredEndpoint,
      });
    }

    return stats;
  }

  getActiveEndpoint(): string | null {
    return this.activeEndpoint;
  }

  getPreferredEndpoint(): string | null {
    return this.preferredEndpoint;
  }

  closeAll(): void {
    this.connections.forEach((connections, endpoint) => {
      connections.forEach(ws => {
        try {
          if (ws.readyState === WebSocket.OPEN) {
            ws.close(1000, 'Pool closing');
          }
        } catch (error) {
          logger.error(`Error closing WebSocket for ${endpoint}:`, error);
        }
      });
    });
    this.connections.clear();
    this.connectionMetrics.clear();
    this.activeEndpoint = null;
  }

  destroy(): void {
    this.isDestroyed = true;

    if (this.healthCheckTimer) {
      clearInterval(this.healthCheckTimer);
      this.healthCheckTimer = null;
    }

    this.closeAll();
    this.endpoints.clear();
    this.healthScores.clear();
    this.preferredEndpoint = null;

    logger.info('ConnectionPool destroyed');
  }

  // ---------------------------------------------------------------------------
  // Private Methods
  // ---------------------------------------------------------------------------

  private selectWeightedRandom(endpoints: Array<[string, EndpointHealth]>): string {
    const totalScore = endpoints.reduce((sum, [_, health]) => sum + health.score, 0);

    if (totalScore === 0) {
      return endpoints[0][0];
    }

    const random = Math.random() * totalScore;
    let accumulated = 0;
    for (const [endpoint, health] of endpoints) {
      accumulated += health.score;
      if (random <= accumulated) {
        return endpoint;
      }
    }

    return endpoints[0][0];
  }

  private selectLeastConnections(endpoints: Array<[string, EndpointHealth]>): string {
    let minConnections = Infinity;
    let bestEndpoint = endpoints[0][0];

    for (const [endpoint] of endpoints) {
      const connections = this.connections.get(endpoint)?.filter(
        ws => ws.readyState === WebSocket.OPEN
      ).length || 0;

      if (connections < minConnections) {
        minConnections = connections;
        bestEndpoint = endpoint;
      }
    }

    return bestEndpoint;
  }

  private selectRoundRobin(endpoints: Array<[string, EndpointHealth]>): string {
    const endpoint = endpoints[this.roundRobinIndex % endpoints.length][0];
    this.roundRobinIndex++;
    return endpoint;
  }

  private updateHealthScore(endpoint: string, isFailure = false): void {
    const health = this.endpoints.get(endpoint);
    if (!health) return;

    let score = health.score;

    if (isFailure) {
      const penalty = Math.min(30, health.consecutiveFailures * 10);
      score = Math.max(this.config.minHealthScore, score - penalty);
    } else {
      const baseIncrease = 5;
      const latencyBonus = this.calculateLatencyBonus(health.averageLatency);
      const consistencyBonus = Math.min(10, health.consecutiveSuccesses * 2);
      score = Math.min(100, score + baseIncrease + latencyBonus + consistencyBonus);
    }

    const timeSinceLastCheck = Date.now() - health.lastChecked;
    if (timeSinceLastCheck > this.config.healthCheckInterval) {
      const decayFactor = 1 - (this.config.scoreDecayRate * timeSinceLastCheck / this.config.healthCheckInterval);
      score = Math.max(this.config.minHealthScore, score * decayFactor);
    }

    health.score = Math.round(score);
    health.lastChecked = Date.now();
  }

  private calculateLatencyBonus(averageLatency: number): number {
    if (averageLatency === 0) return 0;
    if (averageLatency < 50) return 5;
    if (averageLatency < 100) return 3;
    if (averageLatency < 200) return 1;
    return 0;
  }

  private calculateAverageLatency(history: number[]): number {
    if (history.length === 0) return 0;

    let weightedSum = 0;
    let weightSum = 0;

    history.forEach((latency, index) => {
      const weight = index + 1;
      weightedSum += latency * weight;
      weightSum += weight;
    });

    return Math.round(weightedSum / weightSum);
  }

  private getHealthyEndpoints(): Array<[string, EndpointHealth]> {
    return Array.from(this.endpoints.entries())
      .filter(([_, health]) => health.score > this.config.minHealthScore)
      .sort((a, b) => b[1].score - a[1].score);
  }

  private startHealthMonitoring(): void {
    if (this.healthCheckTimer || this.isDestroyed || this.config.healthCheckInterval <= 0) return;

    setTimeout(() => {
      if (!this.isDestroyed) {
        this.performHealthCheck().catch(error => {
          logger.error('Initial health check failed:', error);
        });
      }
    }, 5000);

    this.healthCheckTimer = setInterval(() => {
      if (!this.isDestroyed && !this.healthCheckInProgress) {
        this.performHealthCheck().catch(error => {
          logger.error('Periodic health check failed:', error);
        });
      }
    }, this.config.healthCheckInterval);
  }

  private async performHealthCheck(): Promise<void> {
    if (this.isDestroyed || this.healthCheckInProgress) return;

    this.healthCheckInProgress = true;

    try {
      const promises = [];

      for (const [endpoint, health] of this.endpoints) {
        if (Date.now() - health.lastChecked < this.config.healthCheckInterval / 2) {
          continue;
        }
        promises.push(this.checkEndpointHealth(endpoint));
      }

      await Promise.allSettled(promises);
    } finally {
      this.healthCheckInProgress = false;
    }
  }

  private async checkEndpointHealth(endpoint: string): Promise<void> {
    try {
      const existingConnection = this.connections.get(endpoint)?.find(
        ws => ws.readyState === WebSocket.OPEN
      );

      if (existingConnection) {
        try {
          const start = Date.now();
          existingConnection.send(`PING:${start}`);
          this.recordSuccess(endpoint);
        } catch (error) {
          this.recordFailure(endpoint);
        }
      }
    } catch (error) {
      logger.debug(`Health check for ${endpoint} encountered an error:`, error);
    }
  }

  getDetailedStats(): {
    endpoints: Array<{ url: string; health: any; connections: number; metrics: any[] }>;
    summary: { totalEndpoints: number; healthyEndpoints: number; totalConnections: number; averageHealthScore: number };
  } {
    const endpointStats = [];
    let totalConnections = 0;
    let totalScore = 0;
    let healthyCount = 0;

    for (const [endpoint, health] of this.endpoints) {
      const connections = this.connections.get(endpoint) || [];
      const activeConnections = connections.filter(ws => ws.readyState === WebSocket.OPEN).length;

      endpointStats.push({
        url: endpoint,
        health,
        connections: activeConnections,
        metrics: [],
      });

      totalConnections += activeConnections;
      totalScore += health.score;
      if (health.score > this.config.minHealthScore) {
        healthyCount++;
      }
    }

    return {
      endpoints: endpointStats,
      summary: {
        totalEndpoints: this.endpoints.size,
        healthyEndpoints: healthyCount,
        totalConnections,
        averageHealthScore: this.endpoints.size > 0 ? Math.round(totalScore / this.endpoints.size) : 0,
      },
    };
  }

  setLoadBalancingStrategy(strategy: 'weighted-random' | 'least-connections' | 'round-robin'): void {
    if (!this.isDestroyed) {
      this.config.loadBalancingStrategy = strategy;
      this.roundRobinIndex = 0;
      logger.info(`Load balancing strategy changed to: ${strategy}`);
    }
  }

  setHealthCheckEnabled(enabled: boolean): void {
    if (this.isDestroyed) return;

    this.config.enableHealthCheck = enabled;

    if (enabled && !this.healthCheckTimer && this.config.healthCheckInterval > 0 && this.endpoints.size > 0) {
      this.startHealthMonitoring();
    } else if (!enabled && this.healthCheckTimer) {
      clearInterval(this.healthCheckTimer);
      this.healthCheckTimer = null;
    }
  }
}

File: python-client/wse_client/connection_pool.py
# =============================================================================
# WSE Python Client -- Connection Pool
# =============================================================================
#
# Multi-endpoint management with health scoring and load balancing.
# Port of client/services/ConnectionPool.ts.
# =============================================================================

from __future__ import annotations

import random
import time
from dataclasses import dataclass, field
from typing import Any

from .types import LoadBalancingStrategy


@dataclass
class EndpointHealth:
    """Health score and stats for a single endpoint."""

    url: str
    score: float = 100.0
    active_connections: int = 0
    total_connections: int = 0
    consecutive_failures: int = 0
    consecutive_successes: int = 0
    total_requests: int = 0
    failed_requests: int = 0
    avg_latency_ms: float = 0.0
    latency_history: list[float] = field(default_factory=list)
    last_failure: float | None = None
    last_success: float | None = None
    last_checked: float = 0.0


class ConnectionPool:
    """Multi-endpoint connection pool with health scoring.

    Strategies:
        - ``weighted-random``:  Probability proportional to health score.
        - ``least-connections``: Pick endpoint with fewest active connections.
        - ``round-robin``:      Cycle through endpoints in order.
    """

    # Scoring constants (match TS client)
    MIN_SCORE = 10.0
    MAX_SCORE = 100.0
    HEALTH_CHECK_INTERVAL = 30.0  # seconds
    SCORE_DECAY_RATE = 0.1
    LATENCY_HISTORY_SIZE = 100

    def __init__(
        self,
        endpoints: list[str],
        strategy: LoadBalancingStrategy = LoadBalancingStrategy.WEIGHTED_RANDOM,
    ) -> None:
        """Initialize the connection pool.

        Args:
            endpoints: WebSocket URLs to load-balance across.
            strategy: Load balancing strategy (default: weighted-random).
        """
        self._strategy = strategy
        self._endpoints: dict[str, EndpointHealth] = {
            url: EndpointHealth(url=url, last_checked=time.monotonic())
            for url in endpoints
        }
        self._rr_index = 0
        self._preferred_endpoint: str | None = None

    @property
    def endpoints(self) -> list[str]:
        return list(self._endpoints.keys())

    def select_endpoint(self) -> str:
        """Pick the best endpoint using the configured strategy."""
        if not self._endpoints:
            raise ValueError("No endpoints configured")

        # Prefer sticky endpoint if set and healthy
        if self._preferred_endpoint:
            ep = self._endpoints.get(self._preferred_endpoint)
            if ep and ep.score > self.MIN_SCORE:
                return self._preferred_endpoint

        if self._strategy == LoadBalancingStrategy.LEAST_CONNECTIONS:
            return self._select_least_connections()
        elif self._strategy == LoadBalancingStrategy.ROUND_ROBIN:
            return self._select_round_robin()
        else:
            return self._select_weighted_random()

    def record_success(self, url: str, latency_ms: float = 0.0) -> None:
        ep = self._endpoints.get(url)
        if not ep:
            return
        ep.last_success = time.monotonic()
        ep.consecutive_successes += 1
        ep.consecutive_failures = 0
        ep.total_requests += 1

        if latency_ms >= 0:
            ep.latency_history.append(latency_ms)
            if len(ep.latency_history) > self.LATENCY_HISTORY_SIZE:
                ep.latency_history.pop(0)
            ep.avg_latency_ms = (
                sum(ep.latency_history) / len(ep.latency_history)
                if ep.latency_history
                else 0.0
            )

        self._update_health_score(url)

    def record_failure(self, url: str) -> None:
        ep = self._endpoints.get(url)
        if not ep:
            return
        ep.last_failure = time.monotonic()
        ep.consecutive_failures += 1
        ep.consecutive_successes = 0
        ep.total_requests += 1
        ep.failed_requests += 1

        self._update_health_score(url, is_failure=True)

    def _update_health_score(self, url: str, *, is_failure: bool = False) -> None:
        """Compute health score matching TS ConnectionPool.updateHealthScore."""
        ep = self._endpoints.get(url)
        if not ep:
            return

        score = ep.score

        if is_failure:
            penalty = min(30.0, ep.consecutive_failures * 10.0)
            score = max(self.MIN_SCORE, score - penalty)
        else:
            base_increase = 5.0
            latency_bonus = self._calculate_latency_bonus(ep.avg_latency_ms)
            consistency_bonus = min(10.0, ep.consecutive_successes * 2.0)
            score = min(self.MAX_SCORE, score + base_increase + latency_bonus + consistency_bonus)

        # Time-based decay
        now = time.monotonic()
        time_since_check = now - ep.last_checked
        if time_since_check > self.HEALTH_CHECK_INTERVAL:
            decay_factor = 1.0 - (self.SCORE_DECAY_RATE * time_since_check / self.HEALTH_CHECK_INTERVAL)
            score = max(self.MIN_SCORE, score * decay_factor)

        ep.score = round(score)
        ep.last_checked = now

    @staticmethod
    def _calculate_latency_bonus(avg_latency: float) -> float:
        """Latency bonus matching TS calculateLatencyBonus."""
        if avg_latency == 0:
            return 0.0
        if avg_latency < 50:
            return 5.0
        if avg_latency < 100:
            return 3.0
        if avg_latency < 200:
            return 1.0
        return 0.0

    def add_connection(self, url: str) -> None:
        ep = self._endpoints.get(url)
        if ep:
            ep.active_connections += 1
            ep.total_connections += 1

    def remove_connection(self, url: str) -> None:
        ep = self._endpoints.get(url)
        if ep and ep.active_connections > 0:
            ep.active_connections -= 1

    def get_stats(self) -> list[dict[str, Any]]:
        return [
            {
                "url": ep.url,
                "score": ep.score,
                "active": ep.active_connections,
                "failures": ep.failed_requests,
                "successes": ep.total_requests - ep.failed_requests,
                "avg_latency_ms": round(ep.avg_latency_ms, 1),
                "consecutive_failures": ep.consecutive_failures,
                "consecutive_successes": ep.consecutive_successes,
            }
            for ep in self._endpoints.values()
        ]

    # -- Strategy implementations ---------------------------------------------

    def _select_weighted_random(self) -> str:
        eps = list(self._endpoints.values())
        total = sum(max(ep.score, 1.0) for ep in eps)
        r = random.random() * total
        cumulative = 0.0
        for ep in eps:
            cumulative += max(ep.score, 1.0)
            if r <= cumulative:
                return ep.url
        return eps[-1].url

    def _select_least_connections(self) -> str:
        return min(self._endpoints.values(), key=lambda ep: ep.active_connections).url

    def _select_round_robin(self) -> str:
        urls = list(self._endpoints.keys())
        url = urls[self._rr_index % len(urls)]
        self._rr_index += 1
        return url

    def has_endpoint(self, url: str) -> bool:
        return url in self._endpoints

    def add_endpoint(self, url: str) -> None:
        if url not in self._endpoints:
            self._endpoints[url] = EndpointHealth(url=url, last_checked=time.monotonic())

    def set_active_endpoint(self, url: str) -> None:
        """Mark an endpoint as the preferred active one."""
        self._preferred_endpoint = url

    def reset(self) -> None:
        """Reset all endpoint scores to default."""
        now = time.monotonic()
        for ep in self._endpoints.values():
            ep.score = self.MAX_SCORE
            ep.consecutive_failures = 0
            ep.consecutive_successes = 0
            ep.total_requests = 0
            ep.failed_requests = 0
            ep.active_connections = 0
            ep.latency_history.clear()
            ep.avg_latency_ms = 0.0
            ep.last_failure = None
            ep.last_success = None
            ep.last_checked = now

    def destroy(self) -> None:
        """Clear all endpoints."""
        self._endpoints.clear()
